leftwm-core/src/models/layout_manager.rs

```diff
@@ -19,40 +19,53 @@ impl Default for LayoutMode {
 pub struct LayoutManager {
     pub mode: LayoutMode,
     pub layouts: Vec<Layout>,
+    pub layouts_per_workspaces: HashMap<i32, Vec<Layout>>,
 }
 
 impl LayoutManager {
     pub fn new(config: &impl Config) -> Self {
+        let layouts_per_workspaces = config
+            .workspaces()
+            .unwrap_or_default()
+            .iter()
+            .filter_map(|w| w.id.map(|id| (id, w.layouts.clone())))
+            .collect();
+
         Self {
             mode: config.layout_mode(),
             layouts: config.layouts(),
+            layouts_per_workspaces,
         }
     }
 
     pub fn new_layout(&self) -> Layout {
         *self.layouts.first().unwrap_or(&Layout::default())
     }
 
-    pub fn next_layout(&self, layout: Layout) -> Layout {
-        let mut index = match self.layouts.iter().position(|&x| x == layout) {
-            Some(x) => x as isize,
-            None => return Layout::default(),
-        } + 1;
-        if index >= self.layouts.len() as isize {
-            index = 0;
-        }
-        self.layouts[index as usize]
+    pub fn next_layout(&self, layout: Layout, workspace: &Workspace) -> Layout {
+        *workspace
+            .id
+            .and_then(|id| self.layouts_per_workspaces.get(&id))
+            .or(Some(&self.layouts))
+            .and_then(|layouts| match layouts.iter().position(|&x| x == layout) {
+                Some(index) if index == layouts.len() - 1 => layouts.first(),
+                Some(index) => layouts.get(index + 1),
+                None => None,
+            })
+            .unwrap_or_default()
```

Comment:
This is quite hard to digest, something like:
```rs
match workspace.id {
    Some(id) if self.layouts_per_workspace.contains_key(&id) => *self.layouts_per_workspaces.get(&id),
    _ => match self.layouts.iter().position(|&l| l == layout) {
        Some(index) if index == layouts.len() - 1 => *layouts.first(),
        Some(index) => *layouts.get(index + 1),
        None => layout,
    }
}
```
Feels a bit easier to read.

And yes I would return the current layout if we can't find the next.

Reply:
This does  not quiet work as the second part of the match need to apply to either the layouts from the workspace, or the global one, 

Keeping the same style however, this work and is I think quiet readable:
```rust
        let layouts = workspace
            .id
            .and_then(|id| self.layouts_per_workspaces.get(&id))
            .unwrap_or(&self.layouts);

        let next = match layouts.iter().position(|&x| x == current) {
            Some(index) if index == layouts.len() - 1 => layouts.first(),
            Some(index) => layouts.get(index + 1),
            None => None,
        };

        *next.unwrap_or(&current)
```

I've renamed `layout` to `current` for clarity